design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+
+  initialize   Initializes the schema version table
+  
+  create [-v <version>] <desc> 
+  
+                 Create a pair of migration files with <desc> as description.
+                 Optional: Use the <version> to specify the version to use.
+   
+                 This will created a pair of empty migration files based
+                 on the next schema version.
+               
+                 i.e.                  
+                   $ ksql-migrations create add_users 
+                   Created V000002__Add_users.sql
+
+  apply ( all | next | until <target> )
+  
+              Migrates a schema to new available schema versions (default: all)
+              
+              If 'all' is specified, then it migrates all newer versions available
+              If 'next' is specified, then it migrates only the next available version
+              If 'until <target>' is specified, then it migrates all available versions before <target>
+  
+  info        Displays information about the current and available migrations
+  
+  baseline    Sets the current schema to the specified version
+  
+              This is useful when production environments already have a schema unversioned. This sets
+              the baseline from where to start applying migrations (i.e. kqsl-migrations baseline 1 'Current State')
+  
+  clean       Cleans the schema metadata objects                
+  
+  validate    Validate applied migrations against local files
+  
+              Compares local files checksum against the current metadata checksums to check for migrations files that have changed.
+              This tells the user that their schema might be not valid against their local files.
+  
+Options
+  -c, --config-file  Specifies a configuration file
+  
+  -d, --define       Define variables for the client session (equivalent to the DEFINE statement).
+  
+  --dry-run          Simulates what migrations will be applied in the cluster. Dry-runs do not execute or simulate each migration
+                     statement. Only displays the schema updates that may take place in the cluster without modifying the cluster.
+ 
+  -h, --help         Shows this help  
+    
+```
+
+### Configurations
+
+The tool will support a configuration file where all details for the server and tool migrations will be set. The following
+are the basic configurations that the tool should support.
+
+Example of `ksql-migrations.properties`
+```
+# Server URL and authentication
+ksql.server.url='http://localhost:8080'
+ksql.username='user1'
+ksql.password='pass1'
+
+# Migrations details
+ksql.migrations.streamName='migration_events'
+ksql.migrations.tableName='schema_version'
+ksql.migrations.createSchemas='true'
+```
+
+Note: The command line options and other configurations will also be defined during the implementation.
+
+## Test plan
+
+- Verify positive and negative forward migrations
+- Verify integration with Github and Jenkins
```

Comment:
do you have a description of how these integrations would look like? is that in-scope here or is that something that each developer would integrate on their own?

Reply:
Naa. I got excited in this plan. There is no integration. This plan is just run the main use case for having a migrations tool. So just a few tests that write a few migrations script, commit on github and automate Jenkins to pull and test. If everything works, then we're going to write a blog about it.